feat(oai refactor): Replace `openai_api` with `entrypoints/openai`

CatherineSue opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

This PR aims to refactor and modernize the OpenAI API implementation in SGLang by removing the legacy openai_api module and consolidating it into the entrypoints/openai structure.

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

### 1. Remove Legacy OpenAI API Module

- Removed `sglang/srt/openai_api` directory (commit 5a7df078)
- Replaced all imports with `sglang/srt/entrypoints/openai`
- Removed batch and file endpoints from `http_server.py`
- Updated OpenAI-compatible endpoints in `http_server.py`
- Removed `openai/api_server.py` and `conftest.py` (commit d6d7351c)
- Consolidated functionality into the new structure

### 2. Introduce Centralized Template Management (commit 7d343747)

- Created `TemplateManager` class to eliminate global template state
- Replaced global `chat_template_name`/`completion_template_name` variables
- Integrated `TemplateManager` into `_GlobalState` with proper dependency injection
- Added proper type hints for `TokenizerManager` and `TemplateManager` parameters
- Extracted and reorganized Jinja template utilities
  - Created `jinja_template_utils.py` with template detection and processing logic (moved from `utils.py`)
- Moved utilities from openai/utils.py to avoid improper dependencies
- Renamed `detect_template_content_format` → `detect_jinja_template_content_format`
- Optimized template content format detection
- Detect format during template loading instead of on every request
- Cache detected format in `TemplateManager.jinja_template_content_format`

### 3. Error Fixes and Improvements

- Fixed validation and error handling (commit b0c940cb)
- Added `validation_exception_handler` to http_server.py. This will enforce `Content-Type: application/json` in the request header, which is an OpenAI standard. It also enable FastAPI to automatically decode payload, bypassing the need to handle it manually in each endpoint.
- Fixed import and parameter errors (commits 90cc976e, 40a5c5a2)
- Fixed `is_multimodal` not found error
- Fixed `enable_thinking` parameter issues

### 4. Code Cleanup and Optimization

- Cleaned up unused imports and code (commit f9add484)
- Removed unused imports in` entrypoints/openai`
- Removed unnecessary tests in `test/srt/openai`
- Updated test timings in `run_suite.py`
- Removed `V1RerankReqInput` duplication between `openai.protocol` and `io_struct`
- Improved logging (commits c3be8638, a7d3b35c)
- Use logger.exception for better error details
- Skip embedding in log_requests for concised logs

### 5. Bug Fixes for Specific Tests

- Fixed streaming response format (commit f9add484)
- Updated `test_openai_server.py` to handle separated usage chunk and finish_reason chunk
- Removed batch test in test_openai_server.py
- **Remove `enable_thinking` in reasoning content handling** (commits e583fd9, [(b22eb4a...2dc0341)](https://github.com/sgl-project/sglang/pull/7351/files/0f4018865092804fa5643ee2a7f82874ea9c695f..2dc0341a4c3e52eb4c3805954ee9fca2e44699b4))
  - The problem is related to the `enable_thinking` parameter and it was used for check for reasoning content parsing: `if reasoning_parser and request.separate_reasoning and enable_thinking:`. See [`adapter.py`](https://github.com/sgl-project/sglang/blob/e30ef368abd3be9293838e133d33715a7515a99e/python/sglang/srt/openai_api/adapter.py#L1406), [`serving_chat.py`](https://github.com/sgl-project/sglang/blob/0e635fa6f97d91cf1ef1acb3dc8def9942dc2876/python/sglang/srt/entrypoints/openai/serving_chat.py#L652)
  - Some models like DeepSeek do not support the `enable_thinking` parameter. 
  - This parameter is only supported by Qwen3 in its [chat_template](https://huggingface.co/Qwen/Qwen3-235B-A22B-FP8?chat_template=default#L86). 
  - Our logic was to set this to `True` unless it is explicitly set through `chat_template_kwargs`. However, this brings confusion to developers as not all models support `enable_thinking`.
  - **Suggestion**:
    - The server should not use `enable_thinking` as a condition for reasoning content parsing since it’s specific to Qwen3’s chat template, not a general OpenAI-compatible parameter.
    - It’s more appropriate to treat it as a sampling parameter, controlled entirely by the chat template.
    - ReasoningParser has `force_thinking=True`, which causes models like Qwen3 to treat even the initial user message as part of the reasoning, leading to unexpected behavior.
    - For models like DeepSeek-R1, the <think> tag was removed in their templates, so `enable_thinking` is irrelevant.
    - This PR changed `force_thinking=True` to `False` in `Qwen3Detector` and fixed the logic in `BaseReasoningFormatDetector` to enable it to handle cases when `force_thinking=False`. 
    - **Test**: currently, `test_reasoning_content` is passed.

<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

CatherineSue committed:
fix finish_reason type for protocol.py

CatherineSue committed:
Remove sglang/srt/openai_api

- Replace all imports with sglang/srt/entrypoints/openai
- Remove batch and file endpoints in http_server.py
- Update openai-compatible endpoints in http_server.py

CatherineSue committed:
Remove openai/api_server.py and conftest.py

CatherineSue committed:
Fix chat_template_name not found error

- Add TODO find a better way for the chat template management

CatherineSue committed:
Remove batch test in test_openai_server.py

CatherineSue committed:
Use logger.exception in OpenAIServingBase

- logger.error won't print detailed exec details

CatherineSue committed:
Fix variables not found error in chat_template

- prompt and prompt_ids are not created before referenced
- modalities data should be None in `_apply_jinja_template` otherwise there will be an error in GenerateReqInput.normalize

CatherineSue committed:
refactor: centralize template management and optimize Jinja template utilities

* Create centralized TemplateManager class to eliminate global template state
  - Replace global chat_template_name/completion_template_name variables
  - Integrate TemplateManager into _GlobalState and inject into serving classes
  - Add proper type hints for TokenizerManager and TemplateManager parameters

* Extract and reorganize Jinja template utilities for better separation of concerns
  - Create jinja_template_utils.py with template detection and processing logic
  - Move utilities from openai/utils.py to avoid improper dependencies
  - Rename detect_template_content_format -> detect_jinja_template_content_format

* Optimize template content format detection for better performance
  - Detect format during template loading instead of on every request
  - Cache detected format in TemplateManager.jinja_template_content_format property
  - Add logging for template format detection results

* Clean up codebase and improve maintainability
  - Remove unused imports and clean up import organization
  - Simplify TemplateManager interface by removing unused is_initialized property
  - Update all serving classes (chat, completions, embedding) to use dependency injection
  - Improve code organization and eliminate architectural debt

Benefits:
- ✅ Eliminates global state pollution
- ✅ Better separation of concerns (generic vs OpenAI-specific utilities)
- ✅ Improved performance through caching
- ✅ Cleaner dependency injection pattern
- ✅ More testable and maintainable architecture

CatherineSue committed:
Fix imports, is_multimodal not found, and enable_thinking parameter

jhinpan committed:
Fix test_embedding_openai_server

woodx9 commented on python/sglang/srt/entrypoints/openai/serving_rerank.py:
```diff
@@ -0,0 +1,96 @@
+            if not request.query.strip():
+                return "Query cannot be empty or whitespace only"
+
+        if not request.documents:
```
also need to check none of the document is empty

woodx9 replied:
true, it was due to my carelessness. I will cover this.

> I saw here it only checked documents is not empty: https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/openai_api/adapter.py#L2028-L2043
> 
> Do you mean we need to check
> 
> for doc in request.documents: if not doc.strip(): return "Doc cannot be empty"?

true, it was due to my carelessness. I will cover this.

CatherineSue committed:
Temporarily disable `_generate_request_id_base`

jhinpan committed:
Revert changes in Tokenizer manager & io_struct

CatherineSue committed:
Revert changes in `test_embedding_openai_server.py`

- Test passed after revert

CatherineSue committed:
Add validation_exception_handler to http_server.py

CatherineSue committed:
Skip embedding in log_requests

CatherineSue committed:
Remove V1RerankReqInput from io_struct

- It has a duplication in both openai.protocol and io_struct

CatherineSue committed:
Fix lint and enable_thinking parameter

CatherineSue committed:
Fix typings

CatherineSue committed:
Clear unused imports in entrypoints/openai and update tests

- Clear unused imports in entrypoints/openai
- Update test/srt/openai time in run_suite.py
- Rename `test_openai_adapter.py` to `test_jinja_template_utils.py`
- Remove unnecessary tests in  test/srt/openai
- Update test_openai_server.py since we now separate usage chunk and finish_reason chunk

CatherineSue commented on python/sglang/srt/entrypoints/openai/serving_chat.py:
```diff
@@ -688,9 +683,10 @@ def _build_chat_response(
+            if request.chat_template_kwargs is not None:
+                enable_thinking = request.chat_template_kwargs.get(
+                    "enable_thinking", True
+                )
```
The original issue is that the default is None, it wil override `{}` and raise error on `None.get()`

CatherineSue committed:
Fix notebook curl command and add validation for json media-type in http_server

- Add validate_json_request() function to ensure Content-Type is application/json
- Apply validation dependency to all /v1/ POST endpoints:
  - /v1/completions
  - /v1/chat/completions
  - /v1/embeddings
  - /v1/rerank
  - /v1/score

CatherineSue committed:
Clear notebook output

CatherineSue committed:
Fix rebase error

CatherineSue commented:
Execute Notebook is expeted to fail. It depends on #7400

zhyncs committed:
Merge branch 'main' into chang/remove-adapter

CatherineSue committed:
Fix hidden_states CI

zhyncs committed:
Merge branch 'main' into chang/remove-adapter

CatherineSue committed:
Remove single List[str] input handling in serving_embedding.py

- The handling for a single string in a list can be removed as #7396 is merged.
- Add UT cases in test_openai_server for such case

CatherineSue committed:
Fix reasoning_content CI by removing `enable_thinking` in reasoning content

- Remove enable_thinking in the check condition when handling reasoning content:

enable_thinking is a flag that only supported by Qwen3, in its chat_template. We pass by this parameter in `self.tokenizer_manager.tokenizer.apply_chat_template`.

When handling reasoning content, as other models don't support enable_thinking, this flag should be removed from the check condition.

- Add back `_get_enable_thinking_from_request` as a util function as some reasoning related parser or backend may need it in the future.

woodx9 commented:
I test endpoint embedding rerank and score, all seems good.

CatherineSue committed:
Fix func name in `test_jinja_template_utils.py`

- Correct name should be `detect_jinja_template_content_format`

CatherineSue committed:
Fix Qwen3ReasoningParser to work with enable_thinking parameter

Qwen3 should not assume it is always in reasoning mode. As Qwen3 supports a parameter called `enable_thinking=False` in its chat_template. In this case, it won't generate thinking content.

CatherineSue committed:
Fix BaseReasoningFormatDetector to handle streaming when force_reasoning is False

- Introduce a self._in_reasoning to handle case when force_reasoning is False

CatherineSue committed:
Simplify the logic to handle streaming in BaseReasoningFormatDetector

We don't need to separate _in_reasoning and _force_in_reasoning

CatherineSue committed:
Fix wrong import in jinja_template_utils.py

CatherineSue committed:
Fix tool call start check in BaseFormatDetector

- If there is already a tool, the next token can be simply a tool_call_separator, before following with a `{`

CatherineSue committed:
Merge remote-tracking branch 'origin/main' into moirai/chang/remove-adapter

CatherineSue committed:
Merge branch 'main' into chang/remove-adapter

CatherineSue committed:
Add support for v1/models/{model_id}

- Add `retrieve_model` in `http_server.py`. It will retreive a model's information. And returns 404 if model is not in the served model list.
- Add UT

taegeonum commented:
@CatherineSue Hello, may I ask why files/batch apis have been removed?
